### PR TITLE
lint install

### PR DESCRIPTION
**Pull request by zen0bit**
Not sure where variable arch-apps is defined. Don't found it...

Bug?

### CODE REVIEW

**Comment by ivan-hc** [comment]
`$arch`, this can manage multiple architectures

**Comment by nazdridoy** [comment]
yeah, but does it need to here in the install script. users should get the list that is ralevent to their system.

**Comment by ivan-hc** [comment]
> yeah, but does it need to? here in the install script? users should get the list that is ralevent to their system.

Yes, it is needed, I use AM also on 32bit systems I have installed

**Comment by nazdridoy** [comment]
> currently, at install time it doesn't work. arch is not defined anywhere in the install script

then you have to set the value of 'arch' before you can call 'wget' to download the list. Currently, at install time, it doesn't work because 'arch' is not defined anywhere in the install script.

**Comment by ivan-hc** [comment]
done https://github.com/ivan-hc/AM/commit/c3109884715dd619e032c35fde253d39311f79d7

**Comment by nazdridoy** [comment]
> EDIT: this is better [420178e](https://github.com/ivan-hc/AM/commit/420178e1ad82d888244e823521f836f6a52f3aa8)

yeah, that's what I have suggested.

**Comment by zen0bit** [comment]
Why not merged this?

**Comment by ivan-hc** [comment]
It says there are conflicts

File: INSTALL [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 # CREATING THE MAIN FOLDER FOR "AM"
 mkdir /opt/am /opt/am/.cache /opt/am/modules
-cd /opt/am
+cd /opt/am || echo "/opt/am not exist Exiting!" && exit 1
```
Comment (zen0bit) [inline INSTALL:5]: This should never happend, but just in case!

File: INSTALL [hunk 2]
```diff
@@ -19,15 +19,15 @@ ln -s /opt/am/APP-MANAGER /usr/local/bin/am
...
 arch=$(uname -m)
-wget -q https://raw.githubusercontent.com/ivan-hc/AM-application-manager/main/programs/$arch-apps
+wget -q "https://raw.githubusercontent.com/ivan-hc/AM-application-manager/main/programs/$arch-apps"
 
 # DOWNLOAD MODULES
 MODULES=$(wget -q https://api.github.com/repos/ivan-hc/AM/contents/modules -O - | grep download_url | cut -d '"' -f 4)
 for module in $MODULES; do
 	for v in $module; do
-		cd /opt/am/modules
+		cd /opt/am/modules || echo "/opt/am/modules not exist Exiting!" && exit 2
```
Comment (zen0bit) [inline INSTALL:28]: This should never happend, but just in case!

File: INSTALL [hunk 3]
```diff
@@ -19,15 +19,15 @@ ln -s /opt/am/APP-MANAGER /usr/local/bin/am
...
 		mkdir tmp
-		cd tmp
+		cd tmp || echo "tmp not exist Exiting!" && exit 3
```
Comment (zen0bit) [inline INSTALL:30]: This should never happend, but just in case!

File: INSTALL [hunk 4]
```diff
@@ -38,14 +38,14 @@ done
 
 # ENABLE NON-ROOT PERMISSIONS TO THE MAIN FOLDER FOR THE CURRENT USER
 currentuser=$(who | awk '{print $1}')
-chown -R $currentuser /opt/am 2> /dev/null
+chown -R "$currentuser" /opt/am 2> /dev/null
```
Comment (zen0bit) [inline INSTALL:41]: This never should be problem, but just in case!

File: INSTALL [hunk 5]
```diff
@@ -38,14 +38,14 @@ done
...
 
 # ADD THE BASH COMPLETION SCRIPT
 echo '#!/usr/bin/env bash' >> /opt/am/am-completion.sh; echo 'complete -W "$(cat /opt/am/list 2>/dev/null)" am' >> /opt/am/am-completion.sh
 chmod a+x /opt/am/am-completion.sh
 rm -R -f /opt/am/remove; echo '#!/bin/sh' >> /opt/am/remove
 echo 'rm -R -f /usr/local/bin/am /opt/am /etc/bash_completion.d/am-completion.sh' >> /opt/am/remove
-chmod a+x /opt/am/remove; chown -R $currentuser /opt/am/remove 2>/dev/null 
+chmod a+x /opt/am/remove; chown -R "$currentuser" /opt/am/remove 2>/dev/null
```
Comment (zen0bit) [inline INSTALL:48]: This never should be problem, but just in case!

**Review by zen0bit — COMMENTED** [review]
done
 I think...

**Comment by zen0bit** [comment]
If you merge this and not revert anything I missed, I am done with linting INSTALL
🤝

**Comment by ivan-hc** [comment]
Why you put these exit messages? Are useless:
```
ivan@debian:~$ am -R am
[sudo] password di ivan: 
 ◆ "AM" HAS BEEN SUCCESSFULLY REMOVED!
ivan@debian:~$ sudo '/home/ivan/Scrivania/INSTALL' 
ivan@debian:~$ am -R am
bash: /usr/local/bin/am: File o directory non esistente
```
this is why I say that you should test it on your own device before you do a PR, I'm still unable to remove AM!

**Comment by zen0bit** [comment]
shellcheck

**Comment by zen0bit** [comment]
Just try be prepared that any command can false. So better know where bug belongs...
(Don't do exactly that, but you se the point?)

**Comment by zen0bit** [comment]
Why you can't remove am?

**Comment by ivan-hc** [comment]
https://github.com/ivan-hc/AM/pull/142#issuecomment-1942999278

> Why you can't remove am?

Why don't you stop playng with this script? Today I've had enough of this PR that doesn't work!

You're just wasting my time! It's not possible that you first present them to me as finished and correct and then everything breaks! In the last one I wasn't even able to read the /opt/am/list file

If you see all the commits I made on the INSTALL script today, it's a shame! 2 out of 10 were good! So please! Stop playing with this script! I'm getting really pissed off!

I don't doubt that you want to contribute to the code and do your best, if they were things that work I'd be happy, and I always hope that you do a decent PR... but you have to stop making trouble! Please!

**Comment by zen0bit** [comment]
As you wish
If you find something usefull in my fork, feel free to grab it.

Buona giornata

**Comment by ivan-hc** [comment]
@zen0bit don't be offended, I appreciate the fact that you want to contribute in some way, and I'm sorry that your attempts don't work, I know how much you are worth and how much you have contributed to other projects, it honors you. And I get pissed off when I see that something as simple as "create a folder with `mkdir`" or "create a file with `cat`" ends up breaking everything.

I am the creator of this project, such an important script like "INSTALL" doesn't work for me, do you know what this means? That at that moment, perhaps, someone else downloaded the script and the installation broke.
Result? "AM" is shit! So I, in a hurry, have to correct that damage. And I gave you full trust, and you know it!

Can I not get pissed off about this?

I'm a very impulsive person, I admit, but I think you would be pissed too if I made a PR on your project by breaking everything. There are no saints here.

**Comment by ivan-hc** [comment]
@zen0bit I have not seen your previous comment edited

> shellcheck.net
or install locally (could be integrated into a lot of editors)
Which editor are you using?

Mousepad, the text editor of XFCE Desktop Environment. I don't think it can use that.

In my experience, shellcheck.net is a sh*t.

There is a huge difference between theory and practice. Shellcheck is theory, Mousepad is practice. A war video game is theory, living under bombs is practice. Pornhub is theory, having a partner is practice. Understood?

**Comment by zen0bit** [comment]
> I am the creator of this project, such an important script like "INSTALL" doesn't work for me, do you know what this means? That at that moment, perhaps, someone else downloaded the script and the installation broke. Result? "AM" is shit! So I, in a hurry, have to correct that damage. And I gave you full trust, and you know it!
> 
> Can I not get pissed off about this?

No, merge into `dev` branch first for testing! After will be FULLY tested merge into `main` branch!

Or will be better if install script gets latest release not raw file from github. (common practice)
`wget "https://github.com/ivan-hc/AM/archive/refs/tags/6.tar.gz"`

And use that instead

So two solutions, pick what you want

> I'm a very impulsive person, I admit, but I think you would be pissed too if I made a PR on your project by breaking everything. There are no saints here.

I should expect that, You are 🇮🇹 after all... 😋

**Comment by zen0bit** [comment]
> Mousepad, the text editor of XFCE Desktop Environment. I don't think it can use that.

It's nice simple editor.

> In my experience, shellcheck.net is a sh*t.

excellent tool in hand of experienced one

> There is a huge difference between theory and practice. Shellcheck is theory, Mousepad is practice. A war video game is theory, living under bombs is practice. Pornhub is theory, having a partner is practice. Understood?

And there is huge difference, how you use tool provided. Imagine that you never touch the knife. What possible can happen if you start handling it. Understood?

PS: I am also still learning so mistakes, forgotten peaces are common, but I am trying do better daily. I am trying to contribute to every useful project which i meet. For learning purposes, practice English, human interaction, code, maintenance, and how to treat PRs which I am missing.
😰
And because all bigger projects have some kind of code checking is best learn this practice...
shellcheck/shfmt/other

**Comment by zen0bit** [comment]
My PR to hyfetch with AM support was merged

ehm ehm
Look [here](https://github.com/hykilpikonna/hyfetch/pull/234/checks)

**Comment by nazdridoy** [comment]
# DAMN! ☠️
## Buffet of controversy! 😅

>Think of the author of Thorium Browser

>As if the ones he published on the site were all official

>making himself unpleasant in all projects, see OBS Studio

**Comment by zen0bit** [comment]
Linting is like translator.
it will make your bash same as my without different accent or dialect.

And can be easily repaired, but if you don't use linting everywhere, you will still see shellcheck/user complains.

### Linting is way to cooperate on code most flawesly

**Comment by nazdridoy** [comment]
> PPS: remember my AppMan [issue](https://github.com/ivan-hc/AppMan/issues/11)

i do 😛

```bash
	function validate_directory() {
		local location=$1
		if [[ "$location" == "" ]]; then
			echo -e "--------------------------------------------------------------------------"
			echo "OPERATION ABORTED!"
			exit  1
		elif [[ "$location" == ".local/bin" ]]; then
			echo -e "--------------------------------------------------------------------------\n"
			echo -e " 💀 ERROR, you can't install applications into a '$PATH'\n\n The directory ~/.local/bin and 'bin' folders are used for executables.\n\n Please, choose a different path and retry! \n"
			echo -e "--------------------------------------------------------------------------\n"
			exit  1
			else
			local location_appman="${location/#$HOME}"
			location_appman="${location_appman// /-}"
			location_appman="${location_appman#/}"
			location_appman="${location_appman%/}"
			echo "$location_appman"
		fi
	}

```

and here i was thinking.. "why am i writing a whole routeen to check `.local/bin` ? now i know! :P

By the way, @zen0bit, I do agree that crazy things can happen, and we should be able to handle such cases. But let's test our PRs before we push. For example, I made a bit of complex but 10-character changes in how an install generates an uninstall script, but because of a missing variable, I ended up with `from am -r <package>` to ' rm -rf /opt'. Luckily, I was testing in a temporary Docker container (always do when I see 'rm -rf', especially 'sudo rm -rf').
As for you, @ivan-hc, I would suggest you use Docker or other non-destructive environments to test PRs. And maybe we all should all push to the 'dev' branch. After a PR has at least one review, then you can merge changes into the main branch. AM is mostly feature complete, so not all PRs should be merged right away. Let's review our PRs and then merge

**Comment by zen0bit** [comment]
> and here i was thinking.. "why am i writing a whole routeen to check .local/bin ? now i know! :P

Thank you in advance...

> By the way, @zen0bit, I do agree that crazy things can happen, and we should be able to handle such cases. But let's test our PRs before we push.

Yes I am guilty.

PS: Just let me @nazdridoy  make him agree with Linting, so I can push MegaLinter here... It will repair or spit at me my bugs...

> For example, I made a bit of complex but 10-character changes in how an install generates an uninstall script, but because of a missing variable, I ended up with `from am -r <package>` to ' rm -rf /opt'.

Are you kidding me?

> Luckily, I was testing in a temporary Docker container (always do when I see 'rm -rf', especially 'sudo rm -rf').

I was not that happy today (/opt deleted from production machine)
But nothing important there (Or just not spoted yet) 😁 But everyhing looks like good. Don't know what programs was in /opt will see

> And maybe we all should all push to the 'dev' branch. After a PR has at least one review, then you can merge changes into the main branch. AM is mostly feature complete, so not all PRs should be merged right away. Let's review our PRs and then merge

Advanced same as I suggested, agree

**Comment by nazdridoy** [comment]
> Are you kidding me?

its true. further reviewing i've noticed it could nuke my whole system! `rm -rf /usr/bin /user/share`. regex are scary as fuck! 

> I was not that happy today (/opt deleted from production machine)

backups? I am crazy about my backups.

btrfs snapshoots, BackInTime, 4 external backups, 3 local mirrors, 2 cloud mirrors

**Comment by zen0bit** [comment]
> backups? I am crazy about my backups.
> btrfs snapshoots, BackInTime, 4 external backups, 3 local mirrors, 2 cloud mirrors

All documents and photos sure, multiple drives with partially same data as on other drives (like invoices I have around on 20 places on/offline), dotfiles partialy

Using btrfs but never really used btrfs backup function

Projects yes multiple online git services

maybe I will try snapper or something again, or will create TUI. Since I am lazy as f**k (ex Wind💩ws user) and like make things easier

**Comment by nazdridoy** [comment]
> never really used btrfs backup function

please do. you can never go back not using it. btrfs is probably the only reason i feel Archlinux is stable as Debian (debian users, please don't k☠️ me🙏🏼)

> selected in installator (of course alphabet sorting only) and I stick with it since 

first time hearing someone using btrfs by accident...😅

> (ex Wind💩ws user) 

me too, but WSL made it bearable. However, I snapped when they released Win11 (bloated 💩)

**Comment by nazdridoy** [comment]
marked some comment off topic

**Pull request closed without merging**